### PR TITLE
Implement stream.Writable.toWeb

### LINKED ISSUE #3927: Implement `stream.Writable.toWeb`
### What version of Bun is running?

0.7.2

### What platform is your computer?

_No response_

### What steps can reproduce the bug?

```
console.log('process.stdout', stream.Writable.toWeb(process.stdout))

TypeError: h().newWritableStreamFromStreamWritable is not a function. (In 'h().newWritableStreamFromStreamWritable(X)', 'h().newWritableStreamFromStreamWritable' is undefined)
      at node:stream:2:106081
      at /xxx/src/index.ts:172:32
```

### What is the expected behavior?

_No response_

### What do you see instead?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by TheCactusBlue**
### What does this PR do?

Implements `stream.Writable.toWeb`, resolves #3927

### How did you verify your code works?

Added tests in `test/js/node/stream/node-stream.test.js` (Writable.toWeb) - though I'll probably write more test cases before merging it in

### CODE REVIEW

File: src/js/node/stream.ts [hunk 1]
```diff
@@ -4350,7 +4354,85 @@ var require_writable = __commonJS({
...
+        var { isDestroyed, isWritable } = require_utils();
+        if (isDestroyed(streamWritable) || !isWritable(streamWritable)) {
+          const writable = new WritableStream();
+          writable.close();
+          return writable;
+        }
+        const objectMode = streamWritable.writableObjectMode;
+        const highWaterMark = streamWritable.writableHighWaterMark;
+
+        const evaluateStrategyOrFallback = strategy => {
```
Comment (Jarred-Sumner) [inline src/js/node/stream.ts:4591]: does this need to be a new function each time?

File: test/js/node/stream/node-stream.test.js [hunk 2]
```diff
@@ -386,6 +386,27 @@ it("Readable.fromWeb", async () => {
   expect(Buffer.concat(chunks).toString()).toBe("Hello World!\n");
 });
 
+it("Writable.toWeb", async () => {
```
Comment (Jarred-Sumner) [inline test/js/node/stream/node-stream.test.js]: Where are the tests for Writable.fromWeb?

File: test/js/node/stream/node-stream.test.js [hunk 3]
```diff
@@ -386,6 +386,27 @@ it("Readable.fromWeb", async () => {
...
+  let dataBuffer = "";
+
+  // Create a Node.js Writable stream
+  const nodeWritable = new Writable({
```
Comment (Jarred-Sumner) [inline test/js/node/stream/node-stream.test.js]: Where is error handling tested?

File: src/js/node/stream.ts [hunk 4]
```diff
@@ -5832,6 +5914,21 @@ function getNativeReadableStream(Readable, stream, options) {
   return new NativeReadable(ptr, options);
 }
 
+// Cargo culted from the above function
+// TODO: might not be correct impl
+function getNativeWritableStream(stream, options) {
```
Comment (Jarred-Sumner) [inline src/js/node/stream.ts]: This doesn't refer to a web `WritableStream`, so let's make sure to not name it in a way that makes it sound like it does
```suggestion
function getNativeWritable(stream, options) {
```

File: test/js/node/stream/node-stream.test.js [hunk 5]
```diff
@@ -386,6 +386,27 @@ it("Readable.fromWeb", async () => {
...
+      callback();
+    },
+  });
+
+  const webWritable = Writable.toWeb(nodeWritable);
+  const writer = webWritable.getWriter();
+
+  // Write some data
+  const encoder = new TextEncoder();
+  await writer.write(encoder.encode("Hello, World!"));
```
Comment (Jarred-Sumner) [inline test/js/node/stream/node-stream.test.js]: This should have tests for multiple chunks, not just one.

File: src/js/node/stream.ts [hunk 6]
```diff
@@ -4350,7 +4354,85 @@ var require_writable = __commonJS({
...
+        let controller;
+
+        return new WritableStream(
+          {
+            start(c) {
+              controller = c;
+            },
+
+            write(chunk) {
+              streamWritable.write(chunk);
```
Comment (Jarred-Sumner) [inline src/js/node/stream.ts]: Where is backpressure signaled?

File: src/js/node/stream.ts [hunk 7]
```diff
@@ -4350,7 +4354,85 @@ var require_writable = __commonJS({
...
+          // native = true,
+        } = options;
+
+        if (encoding !== undefined && !Buffer.isEncoding(encoding)) {
+          // TODO: why does this err in typescript?
+          throw new ERR_INVALID_ARG_VALUE(encoding, "options.encoding");
+        }
+        validateBoolean(objectMode, "options.objectMode");
+
+        const nativeStream = getNativeWritableStream(writableStream, options);
```
Comment (Jarred-Sumner) [inline src/js/node/stream.ts]: What happens if it's not a `NativeWritable`? Such as if it has no `$bunNativePtr` and the developer did `new WritableStream` and then passed that to this function?

File: test/js/node/stream/node-stream.test.js [hunk 8]
```diff
@@ -386,6 +386,27 @@ it("Readable.fromWeb", async () => {
...
+
+  // Create a Node.js Writable stream
+  const nodeWritable = new Writable({
+    write(chunk, encoding, callback) {
+      dataBuffer += chunk.toString();
+      callback();
+    },
+  });
+
+  const webWritable = Writable.toWeb(nodeWritable);
```
Comment (Jarred-Sumner) [inline test/js/node/stream/node-stream.test.js]: This should have tests that verify it throws an appropriate exception when invalid input is passed

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Some comments but a good start

File: src/js/node/stream.ts (hunk 1 above)
Comment (TheCactusBlue) [reply src/js/node/stream.ts:4591]: probably not, but mostly copied over from `newStreamReadableFromReadableStream`, and not sure about performance or correctness changes as a result of extracting the function.

File: src/js/node/stream.ts (hunk 6 above)
Comment (TheCactusBlue) [reply src/js/node/stream.ts]: doesn't passing in a strategy automatically signal backpressure?

Comment (Jarred-Sumner) [reply src/js/node/stream.ts]: How does the Node.js Writable become aware of backpressure?

File: test/js/node/stream/node-stream.test.js (hunk 2 above)
Comment (TheCactusBlue) [reply test/js/node/stream/node-stream.test.js]: added!

File: test/js/node/stream/node-stream.test.js (hunk 3 above)
Comment (TheCactusBlue) [reply test/js/node/stream/node-stream.test.js]: just wrote some tests for it; errors resolve to `undefined`, so going to fix that now

Comment (TheCactusBlue) [reply test/js/node/stream/node-stream.test.js]: done!

File: test/js/node/stream/node-stream.test.js (hunk 5 above)
Comment (TheCactusBlue) [reply test/js/node/stream/node-stream.test.js]: modified.

File: test/js/node/stream/node-stream.test.js (hunk 8 above)
Comment (TheCactusBlue) [reply test/js/node/stream/node-stream.test.js]: tested

File: src/js/node/stream.ts (hunk 7 above)
Comment (TheCactusBlue) [reply src/js/node/stream.ts]: fixed now: `getNativeWritableStream` will return `undefined`, and a regular `Writable` will be returned instead if `nativeStream` is undefined

**Pull request closed without merging**